project/wei0107/README.rst

```diff
@@ -6,52 +6,107 @@ Basic Information
+learn, and not tailored for coursework demonstration of numerical methods
+and parallelization.
 
 This project aims to provide a lightweight toolkit that:
 
-- Implements common filters (average, Gaussian, median, bilateral) using convolution and statistical operators.
-- Supports tone and color transformations with linear and nonlinear numerical methods.
+- Implements common filters (average, Gaussian, median, bilateral) using
+  convolution and statistical operators.
+- Supports tone and color transformations with linear and nonlinear
+  numerical methods.
 - Provides sharpening and visual effects (e.g., emboss, cartoon).
 - Offers geometric transformations such as resize, rotate, and crop.
 - Explores performance through parallelization (multi-threads, SIMD).
 
 Prospective Users
 -----------------
 
-- Students in numerical software development or image processing who need a compact, educational, and performance-aware image processing package.
-- Researchers or hobbyists who want to test photo filtering and enhancement algorithms in a simple environment.
-- Everyday users who want to apply lightweight editing to their photos without heavy software dependencies.
+- Students in numerical software development or image processing who need a
+  compact, educational, and performance-aware image processing package.
+- Researchers or hobbyists who want to test photo filtering and enhancement
+  algorithms in a simple environment.
+- Everyday users who want to apply lightweight editing to their photos
+  without heavy software dependencies.
 
 System Architecture
 -------------------
 
 Workflow:
 
 1. Input: Load image as an array (NumPy in Python, raw buffer in C++).
-2. Processing: Apply selected filters or operations (implemented in C++, optionally parallelized).
+2. Processing: Apply selected filters or operations (implemented in C++,
+   optionally parallelized).
 3. Output: Save the processed image or return it as an array.
 
+Raw Memory Buffer Sketch:
+
+.. code-block:: cpp
+
+   // image.hpp
+   #pragma once
+   #include <vector>
+   #include <cstddef>
+   #include <cstdint>
+
+   namespace pf {
+
+   template <class T>
+   class Image {
+   public:
+     Image() = default;
+
+     Image(int h, int w, int c)
+       : h_(h), w_(w), c_(c),
+         stride_bytes_(w * c * int(sizeof(T))),
+         buffer_(std::size_t(h) * std::size_t(w) * std::size_t(c)) {}
+
+     T* data() noexcept { return buffer_.data(); }
+     const T* data() const noexcept { return buffer_.data(); }
+
+     inline T& at(int y, int x, int c = 0) noexcept {
+       return buffer_[(y * w_ + x) * c_ + c];
+     }
+
+     int width() const noexcept { return w_; }
+     int height() const noexcept { return h_; }
+     int channels() const noexcept { return c_; }
+     int stride_bytes() const noexcept { return stride_bytes_; }
+
+   private:
+     int h_ = 0, w_ = 0, c_ = 1;
+     int stride_bytes_ = 0;
+     std::vector<T> buffer_;
```

Comment:
A downside of using `std::vector` for internal buffer is the inability to share the memory buffer, but it's OK for pseudo code.